krabs/krabs/ut.hpp

```diff
@@ -194,11 +194,28 @@ namespace krabs { namespace details {
         const EVENT_RECORD &record,
         const krabs::trace<krabs::details::ut> &trace)
     {
-        for (auto &provider : trace.providers_) {
+        auto found = false;
+
+        // for manifest providers, EventHeader.ProviderId is the Provider GUID
+        for (auto& provider : trace.providers_) {
             if (record.EventHeader.ProviderId == provider.get().guid_) {
+                found = true;
                 provider.get().on_event(record, trace.context_);
             }
         }
+
+        // for MOF providers, EventHeader.Provider is the *Message* GUID
+        // we need to ask TDH for event information in order to determine the
+        // correct provider to pass this event to
+        if (!found) {
```

Comment:
nit: to reduce nesting, consider doing this:

```
if (found) return;

// for MOF providers, EventHeader.Provider is the *Message* GUID
// we need to ask TDH for event information in order to determine the
// correct provider to pass this event to
auto schema = get_event_schema_from_tdh(record);
....
```

Reply:
We could even eliminate the found variable entirely - and just return directly from within the manifest provider loop?

This would mean that only a single provider could be registered for a given GUID. That seems like a fairly sane assumption to make?
In fact, I suspect that this assumption is already implicit with ETW. i.e. a session can't have two sets of differing flags/levels/native filters for the same GUID.